### DISCUSSION TITLE
Wizard form with FieldArray: How to clear previously-saved state?

### DISCUSSION THREAD

**Discussion by kicksnarehat** [react-hook-form/react-hook-form#10017] [Q&A]
When a user changes a field value that's used to conditionally render another field, the simplest solution seems to be using useForm's "shouldUnregister" option, which clears the relevant current form state, but how can I clear state that's already been saved to a back-end on submission? This feels like a typical wizard task (given the scenario where state is saved on "continue").

The code sandbox below attempts this by using setValue in a useEffect return function (called on unmounting). "shouldUnregister" is unset/false, and the behavior works as intended for regular fields -- but FieldArrays present a problem when trying to remove items. As per the official docs, I'm using an individual useForm hook call per-wizard-step. (Have also tried single-useForm, several shouldUnregister options, etc... without success.)

I believe the setValue "clearing" call is running on useEffect/unmount after FieldArray's "remove" function, which results in re-registering the field and/or mangling data by using a stale index reference (`arrayFieldName.NUMBER.fieldName`).

https://codesandbox.io/s/fervent-wave-3ujlg9

My real-world app is very large with dozens of very large forms, so I'm looking for a programmatic/configurable way to accomplish this.

Thanks

**Comment by leapful** [comment]
@kicksnarehat 

Could you please explain why do you need to perform clear value on unmounting Question component  ?

Without the useEffect for clearing data inside Question component, everything works fine.

Working demo: https://codesandbox.io/s/eager-lake-x5xrw9?file=/src/Question.tsx

**Reply by kicksnarehat** [reply]
@leapful - I want to clear previously-saved state so I don't get incorrect/bad data when a user changes fields that are dependent on eachother.

In your example above, uncheck "showDetails" on the first item either save or advance to the next step. Note that submitted state is now incorrectly showing old/bad values for the "name" and "color" fields.

I hope that makes sense.

[edit:] I'd much prefer to avoid calling clearValue in a useEffect, is there a way to configure RHF to handle this for me?

Thanks for replying!

**Reply by leapful** [reply]
@kicksnarehat 

If `showDetails` should clear dependency fields `name` and `color` then those fields should be set "shouldUnregister: true" on field level. For example: https://codesandbox.io/s/beautiful-dijkstra-yw2o25?file=/src/useWizard.tsx

If you want to clear previous form values that has been saved in back-end side, just keep it simple by re-submit the whole form values on change of target fields or when clicking on "Save" button to your BE, not manually update it on FE.

**Reply by kicksnarehat** [reply]
@leapful

Thanks again for responding, very much appreciated. I need to update -- not replace entirely -- data in the back-end to retain other non-form data and relationships. Here's a simplified example (not using FieldArray, Wizard, or useEffect) illustrating the problem:

https://codesandbox.io/s/icy-thunder-1z19wc

Again, the actual form(s) are very large -- hundreds of questions with lots of nested/dependent FieldArrays, so I'm hoping there's a programmatic/configurable solution.

FWIW, my current work-around involves setting shouldUnregister to false, calling setState(fieldname,null) on unmount, using FieldArray.update instead of FieldArray.remove (where update is responsible for hiding objects from the UI, to avoid the probem in the codesandbox in my initial post), and then pruning empty objects from handleSubmit values. And yes, it feels over-complicated!
